### PR TITLE
rec: implement a way to disallow throtteling of auths

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description

TODO:

- [ ] Add new tests for this functionality
- [x] Allow updating the lists via `rec_control`

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
I did not review the code but some SyncRes unit tests related to the throttle code are failing.

**Comment by rgacogne** [comment]
This PR makes it possible to disable throttling for all forward-recurse servers, but it's all or nothing, and limited to these kind of servers. How about having a map instead, so you could disable throttling for selected servers, based on the IP address (but suffix matching would be nice too), which would make it usable for all servers, even regular auths?

File: pdns/pdns_recursor.cc [hunk 1]
```diff
@@ -4219,6 +4240,9 @@ int main(int argc, char **argv)
     ::arg().set("max-tcp-clients","Maximum number of simultaneous TCP clients")="128";
     ::arg().set("server-down-max-fails","Maximum number of consecutive timeouts (and unreachables) to mark a server as down ( 0 => disabled )")="64";
     ::arg().set("server-down-throttle-time","Number of seconds to throttle all queries to a server after being marked as down")="60";
+    ::arg().set("throttle-forward-recurse", "Throttle servers that are forward to with the RD bit set")="yes";
```
Comment (pieterlexis) [inline pdns/pdns_recursor.cc]: needs to be removed

File: pdns/syncres.cc [hunk 2]
```diff
@@ -2618,7 +2618,7 @@ bool SyncRes::processRecords(const std::string& prefix, const DNSName& qname, co
   return done;
 }
 
-bool SyncRes::doResolveAtThisIP(const std::string& prefix, const DNSName& qname, const QType& qtype, LWResult& lwr, boost::optional<Netmask>& ednsmask, const DNSName& auth, bool const sendRDQuery, const DNSName& nsName, const ComboAddress& remoteIP, bool doTCP, bool* truncated)
+bool SyncRes::doResolveAtThisIP(const std::string& prefix, const DNSName& qname, const QType& qtype, LWResult& lwr, boost::optional<Netmask>& ednsmask, const DNSName& auth, bool const sendRDQuery, const DNSName& nsName, const ComboAddress& remoteIP, bool doTCP, bool* truncated, const bool wasForwarded)
```
Comment (rgacogne) [inline pdns/syncres.cc]: I don't think we need `wasForwarded` anymore.

**Comment by rgacogne** [comment]
I don't really understand why but the interface shows a lot of commits that are not related to this PR?!

**Comment by pieterlexis** [comment]
> I don't really understand why but the interface shows a lot of commits that are not related to this PR?!

rebased on master, looks ok now

File: pdns/dnsname.hh [hunk 3]
```diff
@@ -340,14 +385,33 @@ struct SuffixMatchNode
...
+    std::string ret;
+    bool first = true;
+    for (const auto &n: d_tree.getNodes()) {
+      if (!first) {
+        ret += ", ";
+      }
+      first = false;
+      ret += n.toString();
+    }
+    return ret;
```
Comment (rgacogne) [inline pdns/dnsname.hh]: I'm a bit worried about this operation going from `O(1)` to `O(n)`, especially in dnsdist where someone could trigger this by retrieving the list of rules from the `API` for example. Perhaps we could cache the result of this operation to a mutable member variable and invalidate it whenever `::add()` or `::remove()` is called?

File: pdns/dnsname.hh [hunk 4]
```diff
@@ -286,6 +286,42 @@ struct SuffixMatchTree
...
+      // No subnode found, we're done
+      return;
+    }
+
+    // We have found a child
+    labels.pop_back();
+    if (labels.empty()) {
+      // The child is no longer an endnode
+      child->endNode = false;
+      // TODO clear d_value for this node as d_value = T() would break for types
```
Comment (rgacogne) [inline pdns/dnsname.hh]: I'm not sure that's up to this class to handle that. If we really need to, perhaps we could do something with an optional `boost::optional<T> resetValue=boost::none` parameter and set `d_value = resetValue` when clearing, but that seems brittle.

File: pdns/test-dnsname_cc.cc [hunk 5]
```diff
@@ -558,6 +563,11 @@ BOOST_AUTO_TEST_CASE(test_suffixmatch_tree) {
   BOOST_CHECK_EQUAL(*smt.lookup(examplenet), examplenet);
   BOOST_REQUIRE(smt.lookup(net));
   BOOST_CHECK_EQUAL(*smt.lookup(net), net);
+
+  // Remove .net, and check that example.net remains
+  smt.remove(net);
+  BOOST_CHECK(smt.lookup(net) == nullptr);
+  BOOST_CHECK_EQUAL(*smt.lookup(examplenet), examplenet);
 }
```
Comment (rgacogne) [inline pdns/test-dnsname_cc.cc]: I have a few more tests in a different PR, I'll PR them against this branch.

File: pdns/dnsname.hh (hunk 3 above)
Comment (pieterlexis) [reply pdns/dnsname.hh]: we could store the dnsnames in a set and refresh `d_human` on `::add()` and `::remove()`

Comment (pieterlexis) [reply pdns/dnsname.hh]: Although this means that `::add()` and `::remove()` will be a bit more complex.. I wonder what would be called more often, `toString()` or `::add()`/`::remove()`

File: pdns/dnsname.hh [hunk 6]
```diff
@@ -371,49 +371,75 @@ struct SuffixMatchTree
...
+  public:
+    SuffixMatchNode()
+    {}
+    SuffixMatchTree<bool> d_tree;
+
+    void add(const DNSName& dnsname)
+    {
+      d_tree.add(dnsname, true);
+      d_nodes.insert(dnsname);
+      updateHuman();
```
Comment (rgacogne) [inline pdns/dnsname.hh]: I'm a bit worried that we are now doing an `O(n)` operation on insertion, although that seems fine with the way we currently use `SuffixMatchNode`, except perhaps for the `SuffixMatchNodeRule` in dnsdist where it depends how it's used from `Lua`.

File: pdns/dnsname.hh (hunk 4 above)
Comment (pieterlexis) [reply pdns/dnsname.hh]: :+1:, I'll remove the comment and keep it like this.

**Comment by rgacogne** [comment]
Additional tests PR'ed at https://github.com/pieterlexis/pdns/pull/13

File: pdns/rec_channel_rec.cc [hunk 7]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
   return new string("pong\n");
 }
 
+string getDontThrottleNames() {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: Just a nit but should be `static`.

File: pdns/rec_channel_rec.cc [hunk 8]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
...
+  auto dtn = g_dontThrottleNames.getLocal();
+  return dtn->toString() + "\n";
+}
+
+string getDontThrottleNetmasks() {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: same

File: pdns/rec_channel_rec.cc [hunk 9]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
...
+  auto dtn = g_dontThrottleNetmasks.getLocal();
+  return dtn->toString() + "\n";
+}
+
+template<typename T>
+string addDontThrottleNames(T begin, T end) {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: as well

File: pdns/rec_channel_rec.cc [hunk 10]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
...
+
+  g_dontThrottleNames.setState(dnt);
+
+  ret += " to the list of nameservers that may not be throttled";
+  g_log<<Logger::Info<<ret<<", requested via control channel"<<endl;
+  return ret + "\n";
+}
+
+template<typename T>
+string addDontThrottleNetmasks(T begin, T end) {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: this one too

File: pdns/rec_channel_rec.cc [hunk 11]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
...
+
+  g_dontThrottleNetmasks.setState(dnt);
+
+  ret += " to the list of nameserver netmasks that may not be throttled";
+  g_log<<Logger::Info<<ret<<", requested via control channel"<<endl;
+  return ret + "\n";
+}
+
+template<typename T>
+string clearDontThrottleNames(T begin, T end) {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: almost there ;)

File: pdns/rec_channel_rec.cc [hunk 12]
```diff
@@ -1353,6 +1353,190 @@ static string* nopFunction()
...
+
+  g_dontThrottleNames.setState(dnt);
+
+  ret += " from the list of nameservers that may not be throttled";
+  g_log<<Logger::Info<<ret<<", requested via control channel"<<endl;
+  return ret + "\n";
+}
+
+template<typename T>
+string clearDontThrottleNetmasks(T begin, T end) {
```
Comment (rgacogne) [inline pdns/rec_channel_rec.cc]: Last one!

File: pdns/syncres.cc [hunk 13]
```diff
@@ -2780,7 +2780,12 @@ bool SyncRes::doResolveAtThisIP(const std::string& prefix, const DNSName& qname,
...
     }
 
-    if(resolveret != -2 && !chained) { // don't account for resource limits, they are our own fault
+    auto dontThrottleNames = g_dontThrottleNames.getLocal();
+    auto dontThrottleNetmasks = g_dontThrottleNetmasks.getLocal();
+
+    if(resolveret != -2 && !chained && !(dontThrottleNames->check(nsName) || dontThrottleNetmasks->match(remoteIP))) {
+      // don't account for resource limits, they are our own fault
+      // And don't throttle when the IP address is on the dontThrottleNetmasks list or the name is part of dontThrottleNames
       t_sstorage.nsSpeeds[nsName.empty()? DNSName(remoteIP.toStringWithPort()) : nsName].submit(remoteIP, 1000000, &d_now); // 1 sec
```
Comment (rgacogne) [inline pdns/syncres.cc]: It looks like the server can still be throttle on ServFail, Refused, TC=1 over TCP or a lame answer. Is that expected? If so perhaps we should document that.

Comment (pieterlexis) [reply pdns/syncres.cc]: it is expected, this only disable throttling on timeouts

Comment (pieterlexis) [reply pdns/syncres.cc]: otoh, it might defeat the purpose, I'll make it never throttle

File: pdns/recursordist/docs/settings.rst [hunk 14]
```diff
@@ -257,6 +257,40 @@ Operate in the background.
...
+.. _setting-dont-throttle-names:
+
+``dont-throttle-names``
+----------------------------
+.. versionadded:: 4.2.0
+
+-  Comma separated list of domain-names
+-  Default: (empty)
+
+When an authotitative server does not answer a query or sends a reply the recursor does lot like, it is throttled.
```
Comment (rgacogne) [inline pdns/recursordist/docs/settings.rst]:
```suggestion
When an authoritative server does not answer a query or sends a reply the recursor does not like, it is throttled.
```

File: pdns/recursordist/docs/settings.rst [hunk 15]
```diff
@@ -257,6 +257,40 @@ Operate in the background.
...
+.. _setting-dont-throttle-netmasks:
+
+``dont-throttle-netmasks``
+----------------------------
+.. versionadded:: 4.2.0
+
+-  Comma separated list of netmasks
+-  Default: (empty)
+
+When an authotitative server does not answer a query or sends a reply the recursor does lot like, it is throttled.
```
Comment (rgacogne) [inline pdns/recursordist/docs/settings.rst]:
```suggestion
When an authoritative server does not answer a query or sends a reply the recursor does not like, it is throttled.
```

File: pdns/recursordist/docs/settings.rst [hunk 16]
```diff
@@ -257,6 +257,40 @@ Operate in the background.
...
+.. versionadded:: 4.2.0
+
+-  Comma separated list of domain-names
+-  Default: (empty)
+
+When an authotitative server does not answer a query or sends a reply the recursor does lot like, it is throttled.
+Any servers' name suffix-matching the supplied names will never be throttled.
+
+.. warning::
+  Most servers on the internet do not respond for a good reason (overloaded or unreachable), ``dont-throttle-names`` could make this load on the upstream server even higher, resulting in further service degredation.
```
Comment (rgacogne) [inline pdns/recursordist/docs/settings.rst]:
```suggestion
  Most servers on the internet do not respond for a good reason (overloaded or unreachable), ``dont-throttle-names`` could make this load on the upstream server even higher, resulting in further service degradation.
```

File: pdns/recursordist/docs/settings.rst [hunk 17]
```diff
@@ -257,6 +257,40 @@ Operate in the background.
...
+
+When an authotitative server does not answer a query or sends a reply the recursor does lot like, it is throttled.
+Any servers matching the supplied netmasks will never be throttled.
+
+This can come in handy on lossy networks when forwarding, where the same server is configured multiple times (e.g. with ``forward-zones-recurse=example.com=192.0.2.1;192.0.2.1``).
+By default, the PowerDNS Recursor would throttle the "first" server on a timeout and hence not retry the "second" one.
+In this case, ``dont-throttle-netmasks`` could be set to ``192.0.2.1``.
+
+.. warning::
+  Most servers on the internet do not respond for a good reason (overloaded or unreachable), ``dont-throttle-netmasks`` could make this load on the upstream server even higher, resulting in further service degredation.
```
Comment (rgacogne) [inline pdns/recursordist/docs/settings.rst]:
```suggestion
  Most servers on the internet do not respond for a good reason (overloaded or unreachable), ``dont-throttle-netmasks`` could make this load on the upstream server even higher, resulting in further service degradation.
```

**Pull request merged**